src/IO/SharedThreadPools.h

```diff
@@ -1,48 +1,88 @@
 #pragma once
 
-#include <Common/ThreadPool_fwd.h>
+#include <Common/ThreadPool.h>
+#include <Common/CurrentMetrics.h>
 #include <cstdlib>
 #include <memory>
 
 namespace DB
 {
 
-/*
- * ThreadPool used for the IO.
- */
-class IOThreadPool
+template <typename Derived>
+class StaticThreadPool
 {
-    static std::unique_ptr<ThreadPool> instance;
-
 public:
     static void initialize(size_t max_threads, size_t max_free_threads, size_t queue_size);
+    static void reloadConfiguration(size_t max_threads, size_t max_free_threads, size_t queue_size);
     static ThreadPool & get();
 };
 
 
-/*
- * ThreadPool used for the Backup IO.
- */
-class BackupsIOThreadPool
+/// ThreadPool used for the IO.
+class IOThreadPool : public StaticThreadPool<IOThreadPool>
 {
+    friend class StaticThreadPool;
     static std::unique_ptr<ThreadPool> instance;
+    static CurrentMetrics::Metric threads_metric;
+    static CurrentMetrics::Metric threads_active_metric;
+};
 
-public:
-    static void initialize(size_t max_threads, size_t max_free_threads, size_t queue_size);
-    static ThreadPool & get();
+
+/// ThreadPool used for the Backup IO.
+class BackupsIOThreadPool : public StaticThreadPool<BackupsIOThreadPool>
+{
+    friend class StaticThreadPool;
+    static std::unique_ptr<ThreadPool> instance;
+    static CurrentMetrics::Metric threads_metric;
+    static CurrentMetrics::Metric threads_active_metric;
 };
 
 
-/*
- * ThreadPool used for the loading of Outdated data parts for MergeTree tables.
- */
-class OutdatedPartsLoadingThreadPool
+/// ThreadPool used for the loading of Outdated data parts for MergeTree tables.
+class ActivePartsLoadingThreadPool : public StaticThreadPool<ActivePartsLoadingThreadPool>
 {
+    friend class StaticThreadPool;
     static std::unique_ptr<ThreadPool> instance;
+    static CurrentMetrics::Metric threads_metric;
+    static CurrentMetrics::Metric threads_active_metric;
+};
+
+
+/// ThreadPool used for deleting data parts for MergeTree tables.
+class PartsCleaningThreadPool : public StaticThreadPool<PartsCleaningThreadPool>
+{
+    friend class StaticThreadPool;
+    static std::unique_ptr<ThreadPool> instance;
+    static CurrentMetrics::Metric threads_metric;
+    static CurrentMetrics::Metric threads_active_metric;
+};
+
+
+/// This ThreadPool used for the loading of Outdated data parts for MergeTree tables.
+/// Normally we will just load Outdated data parts concurrently in background, but in
+/// case when we need to synchronously wait for the loading to be finished, we can increase
+/// the number of threads by calling enableTurboMode() :-)
+class OutdatedPartsLoadingThreadPool : public StaticThreadPool<OutdatedPartsLoadingThreadPool>
+{
+    friend class StaticThreadPool;
+    static size_t max_threads_turbo;
+    static size_t max_threads;
+    static bool turbo_mode_enabled;
```

Comment:
Looks like we need a mutex to protect these three variables and maybe a `turbo_mode_enabled` must become a counter to deal with concurrent "turbo-enabled" regions (0=disabled, >0=enabled): enable should increment it and disable decrement.

Reply:
Is static mutex even legal?

Reply:
All this starts looking a little bit strange, but it should work. Another maybe more appropriate approach to the problem of shared pools would be the following. Just have one common class with a turbo mode (yes, for all thread pools, but just do not use it for others) and a few helper functions:
```cpp
StaticThreadPool & getXXXThreadPool() { static StaticThreadPool instance(metric1, metric2, ...); return instance; }
```